### PR TITLE
Switch to round-robin scheduling for merges and mutations by default + add config option

### PR DESCRIPTION
**Pull request by serxa**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Concurrent merges are scheduled using round-robin by default to ensure fair and starvation-free operation. Previously in heavily overloaded shards, big merges could possibly be starved by smaller merges due to the use of strict priority scheduling. Added `background_merges_mutations_scheduling_policy` server config option to select scheduling algorithm (`round_robin` or `shortest_task_first`).

### CODE REVIEW

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
But we need small merges to always have higher priority than big merges - this is the whole point. It is ok if big merges will starve indefinitely.

**Comment by nikitamikhaylov** [comment]
@alexey-milovidov We discussed and came to the conclusion that priority will remain the same and smaller merges will be completed earlier just because they are smaller. Let's consider an example and imagine we have only one thread in a background pool

Merges are represented by rows, and the `@` is one step of it (one block execution)

```
@
@@
@@@@
@@@@@
@@@@@@
@@@@@@
@@@@@@@@
@@@@@@@@@
@@@@@@@@@@
``` 

With priority queue this one thread will executed smaller merge first and in general will execute steps row by row (`->`).
If we switch to round-robin then steps will be executed column by column. Of course it will take longer for smaller merges to complete, but still they will be completed earlier than others. If we calculate and take into account that we might have tasks twice more as threads, then it seems to me that with round robin algorithm smaller merges will execute two times slower in comparison with current implementation, but will provide a strong guarantee for bigger ones to finish eventually.

**Comment by alexey-milovidov** [comment]
Ok. Let's make it switchable in config.

**Comment by nikitamikhaylov** [comment]
@serxa Test failures with thread sanitizer are related to this change.

**Review by alexey-milovidov — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 1]
```diff
@@ -225,10 +327,8 @@ class MergeTreeBackgroundExecutor final : boost::noncopyable
     Poco::Logger * log = &Poco::Logger::get("MergeTreeBackgroundExecutor");
 };
 
-extern template class MergeTreeBackgroundExecutor<MergeMutateRuntimeQueue>;
-extern template class MergeTreeBackgroundExecutor<OrdinaryRuntimeQueue>;
-
-using MergeMutateBackgroundExecutor = MergeTreeBackgroundExecutor<MergeMutateRuntimeQueue>;
-using OrdinaryBackgroundExecutor = MergeTreeBackgroundExecutor<OrdinaryRuntimeQueue>;
+extern template class MergeTreeBackgroundExecutor<RoundRobinRuntimeQueue>;
+extern template class MergeTreeBackgroundExecutor<PriorityRuntimeQueue>;
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:331]: Why does it compile? Line 279 requires `pending.updatePolicy(policy);` where `pending` has type `Queue` (template arg of `MergeTreeBackgroundExecutor`). But `RoundRobinRuntimeQueue` and `PriorityRuntimeQueue` do not have this method.

Comment (tavplubix) [reply src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:331]: Well, seems like clang-14 (I'm still using it locally) does not completely support concepts. Ok, so it's time to upgrade :)

Comment (serxa) [reply src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:331]: All this complications are due to the use of explicit template instantiations. Without it compiler would not even try to instantiate methods that are not called. And concepts (or SFINAE) would not be needed here.

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 2]
```diff
@@ -185,6 +267,18 @@ class MergeTreeBackgroundExecutor final : boost::noncopyable
             pool.scheduleOrThrowOnError([this] { threadFunction(); });
     }
 
+    MergeTreeBackgroundExecutor(
+        String name_,
+        size_t threads_count_,
+        size_t max_tasks_count_,
+        CurrentMetrics::Metric metric_,
+        std::string_view policy)
+        requires requires(Queue queue) { queue.updatePolicy(policy); } // Because we use explicit template instantiation
```
Comment (lgbo-ustc) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:276]: `RoundRobinRuntimeQueue` doesn't have `updatePolicy`, we have a compile failure here

**Pull request merged**